keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1243 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Sidecar containers are a new type of containers that start among the Init
+containers, run through the lifecycle of the Pod and don’t block pod
+termination. Kubelet makes a best effort to keep them alive and running while
+other containers are running.
+
+## Motivation
+
+The concept of sidecar containers has been around since the early days of
+Kubernetes. A clear example is [this Kubernetes blog post](https://kubernetes.io/blog/2015/06/the-distributed-system-toolkit-patterns/#example-1-sidecar-containers)
+from 2015 mentioning the sidecar pattern.
+
+Over the years the sidecar pattern has become more common in applications,
+gained popularity and the use cases are getting more diverse. The current
+Kubernetes primitives handle that well, but they fall short for
+several use cases and force weird work-arounds in the applications.
+
+The next sections expand on what the current problems are. But, to give more
+context, it is important to highlight that some companies are already using a
+fork of Kubernetes with this sidecar functionality added (not all
+implementations are the same, but more than one company has a fork for this).
+
+### Problems: jobs with sidecar containers
+
+Imagine you have a Job with two containers: one which does the main processing
+of the job and the other is just a sidecar facilitating it. This sidecar could
+be a service mesh, a metrics gathering statsd server, etc.
+
+When the main processing finishes, the pod won't terminate until the sidecar
+container finishes too. This is problematic for sidecar containers that run
+continuously.
+
+There is no simple way to handle this on Kubernetes today. There are
+work-arounds for this problem, most of them consist of some form of coupling
+between the containers to add some logic where a container that finishes
+communicates it so other containers can react. But it gets tricky when you have
+more than one sidecar container or want to auto-inject sidecars.
+
+The sidecar will also not be restarted for jobs with `restartPolicy:Never`
+when it was OOM killed, which may render the pod unusable if the sidecar provided secure communication to other containers. The issue gets complicated by the fact that 
+sidecar containers typically have smaller request, making them the first target
+as OOM score adjustment uses the request as a main input for calculation.
+
+### Problems: logging/metrics sidecar
+
+A logging sidecar should start before several other containers, to simplify
```

Comment:
I think we should rewrite this use caseto mention a `log forwarding` instead of a `logging` sidecar. Logging is handled natively by container runtimes today and doesn't have races mentioned here.

Reply:
done